File: warmup/2015-europe-regional/testdata/kitchen/submissions/accepted/tobi.java
import java.util.*;
import java.io.*;

public class tobi {
	public static int R, S, M, D, N, K[] = new int[80], B[] = new int[1005], I[][] = new int[80][1005], l, r;
	public static boolean incompat[][] = new boolean[80][80];
	public static final long MAX = 1_000_000_000_000_000_000L;

	public static boolean mult(long prod, long factor) {
		long neu = prod * factor;
		if (neu / factor == prod) {
			prod = neu;
			return true;
		}
		return false;
	}

	public static void main(String[] main) {
		Scanner sc = new Scanner(System.in);
		R = sc.nextInt();
		S = sc.nextInt();
		M = sc.nextInt();
		D = sc.nextInt();
		N = sc.nextInt();
		for (int i = 0; i < R; i++) B[i] = sc.nextInt();
		for (int i = 0; i < S+M+D; i++) {
			K[i] = sc.nextInt();
			for (int j = 0; j < K[i]; j++) I[i][j] = sc.nextInt();
		}
		for (int i = 0; i < N; i++) {
			l = sc.nextInt();
			r = sc.nextInt();
			incompat[l-1][r-1] = incompat[r-1][l-1] = true;
		}
		long ret = 0;
		boolean overflow = false;
		for (int i = 0; i < S; i++) for (int j = S; j < S + M; j++) for (int k = S + M; k < S + M + D; k++) {
			if (incompat[i][j] || incompat[j][k] || incompat[i][k]) continue;
			HashSet<Integer> ings = new HashSet<>();
			int dishes[] = new int[] {i, j, k};
			for (int l = 0; l < 3; l++) {
				for (int m = 0; m < K[dishes[l]]; m++) ings.add(I[dishes[l]][m]);
			}
			long tmp = 1;
			for (Integer it : ings) {
				if (!mult(tmp, B[it-1])) overflow = true;
				tmp *= B[it-1];
				if (tmp > MAX) overflow = true;
			}
			ret += tmp;
			if (ret > MAX) overflow = true;
		}
		if (overflow) System.out.println("too many");
		else System.out.println(ret);
	}
}

File: warmup/2015-europe-regional/testdata/kitchen/submissions/accepted/tobi.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long int ll;
#define FOR(i,a,b) for (int i=(a); i<(b); i++)
int R, S, M, D, N, K[80], B[1005], I[80][1005], l, r;
bool incompat[80][80];

bool mult(ll &prod, ll factor) {
	ll neu = prod * factor;
	if (neu / factor == prod) {
		prod = neu;
		return true;
	}
	return false;
}

int main() {
	cin >> R >> S >> M >> D >> N;
	FOR(i, 0, R) cin >> B[i];
	FOR(i, 0, S+M+D) {
		cin >> K[i];
		FOR(j, 0, K[i]) cin >> I[i][j];
	}
	FOR(i, 0, N) {
		cin >> l >> r;
		incompat[l-1][r-1] = incompat[r-1][l-1] = true;
	}
	ll ret = 0;
	bool overflow = false;
	FOR(i, 0, S) FOR(j, S, S+M) FOR(k, S+M, S+M+D) {
		if (incompat[i][j] || incompat[j][k] || incompat[i][k]) continue;
		map<int,int> ings;
		int dishes[3] = {i, j, k};
		FOR(l, 0, 3) FOR(m, 0, K[dishes[l]]) ings[I[dishes[l]][m]] = 1;
		ll tmp = 1;
		for (auto it : ings) {
			if (!mult(tmp, B[it.first-1])) overflow = true;
			if (tmp > 1e18L) overflow = true;
		}
		ret += tmp;
		if (ret > 1e18L) overflow = true;
	}
	if (overflow) cout << "too many" << endl;
	else cout << ret << endl;

	return 0;
}
